pkg/engine/executor/project.go

```diff
@@ -0,0 +1,55 @@
+package executor
+
+import (
+	"fmt"
+
+	"github.com/apache/arrow-go/v18/arrow"
+	"github.com/apache/arrow-go/v18/arrow/array"
+
+	"github.com/grafana/loki/v3/pkg/engine/planner/physical"
+)
+
+func NewProjectPipeline(input Pipeline, columns []physical.ColumnExpression, evaluator *expressionEvaluator) (*GenericPipeline, error) {
+	// Get the column names from the projection expressions
+	columnNames := make([]string, len(columns))
+	for i, col := range columns {
+		if colExpr, ok := col.(*physical.ColumnExpr); ok {
+			columnNames[i] = colExpr.Ref.Column
+		} else {
+			return nil, fmt.Errorf("projection column %d is not a column expression", i)
+		}
+	}
+
+	return newGenericPipeline(Local, func(inputs []Pipeline) state {
+		// Pull the next item from the input pipeline
+		input := inputs[0]
+		err := input.Read()
+		if err != nil {
+			return failureState(err)
+		}
+
+		batch, err := input.Value()
```

Comment:
do not have a strong opinion about this. but with this API design, the same err is returned twice once with `input.Read()` and then again with `input.Value()`

Reply:
Can you explain a bit more? The first error comes from the child Pipeline (via Read()) and the second from the same pipeline via Value(). I don't think we can make assumptions about these being the same, although it's possible we could revisit the Pipeline interface, but that's beyond the scope of this PR